### PR TITLE
[#82] Remove maybeT, add diOptional

### PR DESCRIPTION
**Pull request by willbasky**
Resolves: [#82](https://github.com/kowainik/tomland/issues/82)

### CODE REVIEW

File: src/Toml/Bi/Combinators.hs [hunk 1]
```diff
@@ -169,18 +169,13 @@ arrayOf bimap key = Codec input output
...
-maybeT :: forall a . (Key -> TomlCodec a) -> Key -> TomlCodec (Maybe a)
-maybeT converter key = let codec = converter key in Codec
-    { codecRead  = (Just <$> codecRead codec) `catchError` handleNotFound
+diOptional :: (Functor r, Functor w, Applicative w) => Codec r w c a -> Codec r w (Maybe c) (Maybe a)
+diOptional Codec {..} = Codec
+    { codecRead = Just <$> codecRead
     , codecWrite = \case
         Nothing -> pure Nothing
-        Just v  -> codecWrite codec v >> pure (Just v)
+        Just c -> Just <$> codecWrite c
```
Comment (ChShersh) [inline src/Toml/Bi/Combinators.hs]: Nice! You've just implemented `Traversable` instance for `Maybe`. If you look at this instance, you probably might notice how to implement `codecWrite` in more concise way:

* https://hackage.haskell.org/package/base-4.12.0.0/docs/src/Data.Traversable.html#line-232

File: src/Toml/Bi/Combinators.hs [hunk 2]
```diff
@@ -169,18 +169,13 @@ arrayOf bimap key = Codec input output
         a <$ modify (\(TOML vals tables) -> TOML (HashMap.insert key anyVal vals) tables)
 
 -- | Bidirectional converter for @Maybe smth@ values.
-maybeT :: forall a . (Key -> TomlCodec a) -> Key -> TomlCodec (Maybe a)
-maybeT converter key = let codec = converter key in Codec
-    { codecRead  = (Just <$> codecRead codec) `catchError` handleNotFound
+diOptional :: (Functor r, Functor w, Applicative w) => Codec r w c a -> Codec r w (Maybe c) (Maybe a)
+diOptional Codec {..} = Codec
```
Comment (ChShersh) [inline src/Toml/Bi/Combinators.hs]: There should be no space between `Codec` and `{..}`

**Comment by vrom911** [comment]
Travis fails with an error for `tasty-quickcheck` for `Cabal`. Seems like this package has breaking change in the newest version and one of our dependencies (which depends on the `tasty-quickcheck` because we don't 🙂 ) didn't have the upper bounds for that. I will investigate this..

**Comment by vrom911** [comment]
It seems it fixed itself. Version `2.12.6.1` of the `QuickCheck` just a few hours after the `2.12.6`, so guess the issue was in there, cause it works now 🙂

File: src/Toml/Bi/Combinators.hs (hunk 1 above)
Comment (willbasky) [reply src/Toml/Bi/Combinators.hs]: If so, maybe someday I will implement somewhat function for `relude` :)

**Comment by willbasky** [comment]
While I slept the problem was solved :)

**Pull request merged**